rasa/core/nlg/generator.py

```diff
@@ -91,3 +90,121 @@ def _load_from_module_name_in_endpoint_config(
             f"'{endpoint_config.type}'. Failed to create a "
             f"`NaturalLanguageGenerator` instance. Error: {e}"
         )
+
+
+class ResponseVariationFilter:
+    """Filters response variations based on the channel, action and condition."""
+
+    def __init__(self, responses: Dict[Text, List[Dict[Text, Any]]]) -> None:
+        self.responses = responses
+
+    @staticmethod
+    def _matches_filled_slots(
+        filled_slots: Dict[Text, Any], response: Dict[Text, Any]
+    ) -> bool:
+        """Checks if the conditional response variation matches the filled slots."""
+        constraints = response.get(RESPONSE_CONDITION, [])
+        for constraint in constraints:
+            name = constraint["name"]
+            value = constraint["value"]
+            filled_slots_value = filled_slots.get(name)
+            if isinstance(filled_slots_value, str) and isinstance(value, str):
+                if filled_slots_value.casefold() != value.casefold():
+                    return False
+            elif filled_slots_value != value:
+                return False
+
+        return True
+
+    def responses_for_utter_action(
+        self,
+        utter_action: Text,
+        output_channel: Text,
+        filled_slots: Dict[Text, Any],
+    ) -> List[Dict[Text, Any]]:
+        """Returns array of responses that fit the channel, action and condition."""
+        default_responses = list(
+            filter(
+                lambda x: (x.get(RESPONSE_CONDITION) is None),
+                self.responses[utter_action],
+            )
+        )
+        conditional_responses = list(
+            filter(
+                lambda x: (
+                    x.get(RESPONSE_CONDITION)
+                    and self._matches_filled_slots(
+                        filled_slots=filled_slots, response=x
+                    )
+                ),
+                self.responses[utter_action],
+            )
+        )
+
+        conditional_channel = list(
+            filter(lambda x: (x.get(CHANNEL) == output_channel), conditional_responses)
+        )
+        conditional_no_channel = list(
+            filter(lambda x: (x.get(CHANNEL) is None), conditional_responses)
+        )
+        default_channel = list(
+            filter(lambda x: (x.get(CHANNEL) == output_channel), default_responses)
+        )
+        default_no_channel = list(
+            filter(lambda x: (x.get(CHANNEL) is None), default_responses)
+        )
```

Comment:
Wouldn't it be more performant to iterate through the list `conditional_responses` and `default_responses` to filter them? Instead of iterating through them twice

Reply:
As per my reply [here](https://github.com/RasaHQ/rasa/pull/12575/files#r1246257045), refactoring pre-existing code is not in scope.